llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -27924,6 +27924,24 @@ bool AArch64TargetLowering::isIntDivCheap(EVT VT, AttributeList Attr) const {
   return OptSize && !VT.isVector();
 }
 
+bool AArch64TargetLowering::canMergeStoresTo(unsigned AddressSpace, EVT MemVT,
+                                             const MachineFunction &MF) const {
+  // Avoid merging stores into fixed-length vectors when Neon is unavailable.
+  // In future, we could allow this when SVE is available, but currently,
+  // the SVE lowerings for BUILD_VECTOR are limited to a few specific cases (and
+  // the general lowering may introduce stack spills/reloads).
```

Comment:
[Just thinking out loud here] My understanding is that for the example in the test, the reason we don't want to do this optimisation is because we can use the `stp` instructions instead, there is no upside to merging the stores although there is a possible downside that the insert operation is expensive. At the moment, it is expensive because we use a spill/reload, but for streaming[-compatible] SVE we could implement the operation using the SVE `INSR` instruction, which may not be any less efficient than the NEON operation if the value being inserted is also in a FPR/SIMD register. With the lack of upside, disabling the merging of stores avoids this complexity altogether, which understandably is the route chosen here.

I guess the question is; for which cases is merging stores beneficial when NEON is available? and for those cases, can we implement these efficiently using Streaming SVE?

Reply:
I think there are two (slightly) independent cases here. There's unwanted store merging (for non-streaming functions), because we could just use a stp instead. E.g.

```
mov v0.s[1], v1.s[0]
str d0, [x0]
```
->

```
stp s0, s1, [x0]
```

That's not fixed in this PR.

Then there's streaming mode store merging, which results in stack spills due to the BUILD_VECTOR lowering. Disabling store mering means, in some cases, we use a more preferable `stp` in streaming mode, but that's a secondary goal here; the main aim is to avoid the stack spills.

As for a streaming-mode/SVE BUILD_VECTOR lowering, I think there are a few options, but likely not as efficient as NEON (though maybe others have better ideas :smile:).

E.g. for <4 x float>:

You could make a chain of `INSR`:
```
insr    z3.s, s2
insr    z3.s, s1
insr    z3.s, s0
str     q3, [x0]
```

But `INSR` has a higher latency than a `MOV`. Also, there is a dependency chain here, as each `INSR` depends on the previous one.

Another option is a chain of `ZIP1`:

```
zip1    z2.s, z2.s, z3.s
zip1    z0.s, z0.s, z1.s
zip1    z0.d, z0.d, z2.d
str     q0, [x0]
```

This seems like it may be more efficient than `INSR`, and also allows for a shorter dependency chain (logn), but it is still likely not as efficient  as just `MOV`s.

Reply:
I agree, ZIP would indeed be a (much) better choice than INSR. I'm also happy with the intent of this PR. The part that isn't entirely clear to me yet is for which cases we'd want to enable this merging when we _do_ have optimal SVE codegen.

Reply:
FWIW, I just manually tried the <4 x float> case: https://godbolt.org/z/MYzrdahjh 
I'd say that the SVE version using zip1 is no less efficient than this.

Reply:
> The part that isn't entirely clear to me yet is for which cases we'd want to enable this merging when we do have optimal SVE codegen.

I'm not sure either, but the original commit (from way back when) for the DAG combine only handled stores of constants and loads and noted it's generally not profitable (see: https://github.com/llvm/llvm-project/commit/7cbc12a41d72bf106e3075878b4f1f6047085e87). The way it's merging stores of extracts here is a little odd and _maybe_ unintentional? It'd make more sense to do the merge if it was storing lanes from the same vector, which is not the case here.